src/LinAlg/Test/MatrixTests.h

```diff
@@ -0,0 +1,227 @@
+    I(i,i) = 1.0;
+  for (size_t i = 1; i <= 3; ++i)
+    I2(i,i) = 1.0;
+
+  ASSERT_TRUE(A.multiply(u,x,1.0,0.0,false,3,4,1,2));
+  ASSERT_TRUE(A.multiply(v,y,1.0,0.0,true,4,2));
+
+  EXPECT_FLOAT_EQ(x(3),370.0);
+  EXPECT_FLOAT_EQ(x(7),410.0);
+  EXPECT_FLOAT_EQ(x(11),450.0);
+  EXPECT_FLOAT_EQ(y(1),38.0);
+  EXPECT_FLOAT_EQ(y(3),83.0);
+  EXPECT_FLOAT_EQ(y(5),128.0);
+  EXPECT_FLOAT_EQ(y(7),173.0);
+  EXPECT_FLOAT_EQ(y(9),218.0);
+
+  ASSERT_TRUE(A.multiply(u,x,1.0,-1.0,false,3,4,1,2));
+  ASSERT_TRUE(A.multiply(v,y,1.0,-1.0,true,4,2));
+
+  EXPECT_FLOAT_EQ(x.sum(),0.0);
+  EXPECT_FLOAT_EQ(y.sum(),0.0);
+
+  u.std::template vector<Scalar>::resize(5);
+  ASSERT_TRUE(A.multiply(u,v));
+  v *= 0.5;
+
+  EXPECT_FLOAT_EQ(v(1),67.5);
+  EXPECT_FLOAT_EQ(v(2),75.0);
+  EXPECT_FLOAT_EQ(v(3),82.5);
+
+  B.multiplyMat(A, static_cast<const std::vector<Scalar>&>(I));
+  B2.multiplyMat(A, static_cast<const std::vector<Scalar>&>(I), false, true);
+  B3.multiplyMat(A, static_cast<const std::vector<Scalar>&>(I2), true, false);
+  EXPECT_EQ(B.rows(), A.rows());
+  EXPECT_EQ(B.cols(), A.cols());
+  for (size_t i = 1; i <= 3; ++i)
+    for (size_t j = 1; j <= 5; ++j) {
+      EXPECT_FLOAT_EQ(B(i,j), A(i,j));
+      EXPECT_FLOAT_EQ(B2(i,j), 2.0*A(i,j));
+      EXPECT_FLOAT_EQ(B3(j,i), A(i,j));
+    }
+
+  if constexpr (std::is_same_v<Scalar,Real>) {
+    utl::vector<Scalar> v2 = A * u;
+    EXPECT_FLOAT_EQ(v2(1),135.0);
+    EXPECT_FLOAT_EQ(v2(2),150.0);
+    EXPECT_FLOAT_EQ(v2(3),165.0);
+  }
+}
+
+
+template<class Scalar>
+void normTest()
+{
+  utl::matrix<Scalar> a(4,5);
+  std::iota(a.begin(),a.end(),1.0);
+  std::cout <<"A:"<< a;
+
+  EXPECT_FLOAT_EQ(a.sum(),210.0);
+  EXPECT_FLOAT_EQ(a.sum(5),34.0);
+  EXPECT_FLOAT_EQ(a.asum(5),34.0);
+  EXPECT_FLOAT_EQ(a.trace(),34.0);
+  EXPECT_NEAR(a.norm2(5),sqrt(414.0),std::numeric_limits<Scalar>::epsilon()*10.0);
+  EXPECT_FLOAT_EQ(a.normInf(),60.0);
+}
+
+
+template<class Scalar>
+void outerProductTest()
+{
+  constexpr size_t size1 = 3;
+  constexpr size_t size2 = 4;
+
+  utl::vector<Scalar> u(size1);
+  utl::vector<Scalar> v(size2);
+  std::iota(u.begin(), u.end(), 1.0);
+  std::iota(v.begin(), v.end(), 2.0);
+
+  utl::matrix<Scalar> A;
+  A.outer_product(u, v);
+
+  utl::matrix<Scalar> A2(A);
+  A2.outer_product(u, v, true, 2.0);
+
+  EXPECT_EQ(A.rows(), size1);
+  EXPECT_EQ(A.cols(), size2);
+
+  for (size_t i = 1; i <= size1; ++i)
+    for (size_t j = 1; j <= size2; ++j) {
+      EXPECT_FLOAT_EQ(A(i,j), u(i)*v(j));
+      EXPECT_FLOAT_EQ(A2(i,j), 3.0*u(i)*v(j));
+    }
+}
+
+
+template<class Scalar>
+void matrix3DMultiplyTest()
+{
+  std::vector<Scalar> a(10);
+  utl::matrix<Scalar> A(2,5);
+  utl::matrix3d<Scalar> B(5,4,3), C, D;
+
+  std::iota(a.begin(),a.end(),1.0);
+  std::iota(A.begin(),A.end(),1.0);
+  std::iota(B.begin(),B.end(),1.0);
+
+  C.multiply(A,B);
+  ASSERT_TRUE(D.multiplyMat(a,B));
+
+  typename std::vector<Scalar>::const_iterator c = C.begin();
+  for (Scalar d : D)
+    EXPECT_EQ(d,*(c++));
```

Comment:
Should maybe use `EXPECT_FLOAT_EQ` here as well, assuming `Scalar` is either `float` or `double`?